subcommands/waves/rollout.go

```diff
@@ -98,8 +109,52 @@ func doRolloutWave(cmd *cobra.Command, args []string) {
 		Limit:      limit,
 		Percentage: percentage,
 		Uuids:      uuids,
+		DryRun:     dryRun,
+		PrintUuids: printUuids,
+		PrintNames: printNames,
+	}
+	res, err := api.FactoryRolloutWave(factory, wave, options)
+	subcommands.DieNotNil(err)
+
+	if len(group) == 0 {
+		if dryRun {
+			fmt.Printf("A wave would be rolled out to %d devices in your factory.\n", res.DeviceNum)
+		} else {
+			fmt.Printf("A wave is being rolled out to %d devices in your factory.\n", res.DeviceNum)
+		}
+	} else {
+		if dryRun {
+			fmt.Printf("A wave would be rolled out to %d devices in group \"%s\".\n", res.DeviceNum, group)
+			if isFullRollout {
+				fmt.Printf("It would also be rolled out to all new devices in group \"%s\".\n", group)
+			}
+		} else {
+			fmt.Printf("A wave is being rolled out to %d devices in group \"%s\".\n", res.DeviceNum, group)
+			if isFullRollout {
+				fmt.Printf("It will also be rolled out to all new devices in group \"%s\".\n", group)
+			}
+		}
+	}
+	if (printUuids && len(res.DeviceUuids) == 0) || (printNames && len(res.DeviceNames) == 0) {
+		// This can happen when rolling out to a large group without any limit specified.
+		fmt.Println("A full list of these devices exceeds 10000, so it cannot be printed.")
+	} else if printUuids || printNames {
+		fmt.Println("Below is a full list of these devices:")
+		for i := 0; i < res.DeviceNum; i++ {
+			if i > 0 {
+				fmt.Print(", ")
+			}
+			if printUuids {
+				fmt.Print(res.DeviceUuids[i])
+				if printNames {
+					fmt.Printf(" (%s)", res.DeviceNames[i])
+				}
```

Comment:
I missed this earlier - you are handling all 3 cases. I thought about trying to simplify this logic but it wound up using `text/template` and defining a struct. So this is probably *almost* as short - and probably *much* more efficient.

Reply:
Yes, we basically have 3 options:
1. A loop with 3 ifs.
2. 3 ifs, each with its own loop.
3. A loop with a static template containing 3 ifs.

What @mike-sul suggests above is practically equivalent to the template implemented as a set of functions (i.e. option 3).

All 3 choices are technically equivalent, but as per my personal taste the first option is more readable.
From the speed perspective 2 is probably marginally faster than 1, yet both win option 3 by a decent margin.